### ISSUE TITLE
KV opacity can be broken by compaction

### ISSUE THREAD

**Issue by eddyashton** [microsoft/CCF#1615]
If we have a long-running transaction that gets views at different points, it can break opacity if a compaction happens while it is in-flight:

```
    kv::Tx tx;

    auto view_a = tx.get_view(map_a);
    // This transaction does something slow. Meanwhile...

    // ...another transaction commits...
    increment_vals();
    // ...and is compacted...
    kv_store.compact(kv_store.current_version());

    // ...then the original transaction proceeds, expecting to read a single version
    auto view_b = tx.get_view(map_b);

    auto a_opt = view_a->get(key_a);
    auto b_opt = view_b->get(key_b);
    
    REQUIRE(a_opt == b_opt);
```

The call to `get_view(map_b)` eventually calls `create_view(Version version)` in `untyped_map.h`, using the `read_version` assigned in this `tx` for the `get_view` on `map_a)`. But this version is _older_ than anything in `map_b`'s roll - the requested version was deleted during compact. We currently have an odd case at the end of `create_view` which returns _a_ view, but this is definitely wrong. We're not sure why this is ever valid, so should be removed.

Fixes:
- If we can't find a suitable version in the roll during `create_view`, return nullptr or an error. Somewhere this becomes an error which propagates to the caller (past the transaction logic) to retry the entire transaction.
- Explore extending the range of states which are kept in the roll at compaction. Perhaps we can track exactly which versions are still potentially requestable (a pending transaction has got them as `read_version`), but we still need an upper limit on the length of the roll for memory growth.

**Comment by achamayou** [comment]
Raising in create_view() when a suitable version can't be found seems like the right thing to do.

The second bullet point makes perfect sense, but is an optimisation rather than a fix, I think.

**Comment by eddyashton** [comment]
Closing this issue as resolved in #1817. Once we support multi-threaded execution and have long-running, potentially asynchronous transactions, we may revisit the optimisation mentioned in the second bullet point, but it is not a priority currently.
